### ISSUE TITLE
Inline custom widgets 

### ISSUE THREAD

**Issue by ghostbear** [netlify/netlify-cms#5065]
Hi!

I'm experiencing some issues with creating custom widgets that are inline with the text, like this
```html
Lorem ipsum dolor sit amet, <Navigation item="more"/> consectetur adipiscing elit.
```
I'm expecting this to show a selector that writers can use to select a navigation item when using rich text, but none is shown and the preview text is also not shown.

This works fine when they are on their own, like this
```html
Lorem ipsum dolor sit amet,

<Navigation item="more"/>

consectetur adipiscing elit.
```
This shows a selector that the writers can use to select a navigation item when using rich text

So is it possible to even get it to show custom widgets when they are inline, like **bold** or _italic_? Or am I trying to do something that is not possible to do with custom widgets currently?

---

```js
CMS.registerEditorComponent({
	id: "navigation",
	label: "Navigation",
	fields: [
		{
			name: "item",
			label: "Item",
			widget: "select",
			options: Object.keys(itemMap).map(key => {
				return { label: itemMap[key].text, value: key }
			}),
			required: true
		},
	],
	pattern: /<Navigation item="([a-zA-Z\-_]+)"\/>/,
	fromBlock: function(match) {
		return {
			item: match[1] || 'library',
		};
	},
	toBlock: function(obj) {
		return `<Navigation item="${obj.item || 'library'}"/>`;
	},
	toPreview: function(obj) {	
		return `<p>${obj.item || 'library'}<p/>`
	},
});
```
This code is the code for the custom widget. It gives a warning on all items that are inline for example `Sent invalid data to remark. Plugin: navigation. Value: <Navigation item="more"/>. Data: {"item":"more"}`. I have tried looking for a solution but I have not found one.

**Comment by doompadee** [comment]
Same issue here. I'm trying to integrate custom Hugo shortcodes with the editor, but it currently does not seem possible which makes Netlify less ideal if non tech-savvy writers are involved.

**Comment by rudasn** [comment]
Anybody had any luck with getting inline custom editor widgets to work?

I'm also looking into using the `rel` or `refrel` Hugo shortcode into the Richtext editor to allow users to add links to other parts of the site using the shortcode (as opposed to copy & pasting a URL). Any links created with this widget should be displayed inline (not block) in the Richtext view (as well as the end result, obviously). The widget would allow to select a page to link to, specify the title, and other properties like target).

From my understanding, that would require a way to specify to netlify that the custom widget (let's call it the Hugo Links widget) should be rendered as an inline element in the richtext editor (using a text slate object as opposed to a block).

Then, using a customised version of the existing Object widget we could display the HugoLinks widget inline with the text, click on it to expand/edit the properties, and click again to toggle back to inline.

I just started to look into netlify a couple of days ago so this may not be right approach, so I would greatly appreciate the input of more experienced members. The ability to not easily handle links is kind of a deal breaker, especially for non-technical users.

**Comment by HenkVanMaanen** [comment]
Any update on this? It's weird that you can match on inline patterns but not create an inline component.

**Comment by tpluscode** [comment]
Trying now with decap 3.7.0, I too fail to create a custom editor widget. 

It fails in `remark-parser/lib/tokenizer.js` with an error `Incorrectly eaten value: ...`. The custom widget plugin correctly matches my block and I see correct `subvalue` being passed to `validateEat(subvalue)` but it fails on this line: https://github.com/remarkjs/remark/blob/remark-parse%406.0.3/packages/remark-parse/lib/tokenizer.js#L146

The comparison fails because you can see `substring` starting from the document's beginning and not where the match occurs...

I see you're using an old version of remark. Maybe it should be updated?

**Comment by tpluscode** [comment]
I think I found the problem in the shortcodes plugin. See the first comment in `tokenizeShortcode`: 
https://github.com/decaporg/decap-cms/blob/3e126b3addf02dc71644bc4d7cd20bfb8062fa48/packages/decap-cms-widget-markdown/src/serializers/remarkShortcodes.js#L47-L67

tl;dr; the plugin matches too greedily, ignoring content before the widget's pattern. In other words, the plugins matches a shortcode prematurely and then has remark try to consume it before actually consuming the preceding content first.

Here's an example, using the youtube block from the decap repo. Given markdown like

```md
bar


{{< youtube bar >}}
```

First, it will match `bar` to a paragraph block. All good so far
Then the rest gets matched to `youtube` (including the blank lines). But the actual match is only the `{{< youtube bar >}}` line. This trips remark because the blank lines shoud be matched first to `blankLine`.

My initial reflex is to remove `trimmedLines` and only use the regex in line `63` but it is there for a reason, so a more robust handling of those patterns would be needed. Anyway, I can start a PR and see where we go from there.
